### PR TITLE
Cover generators with tests

### PR DESCRIPTION
**Pull request by KonstantinLapkovsky**
Issue #49

### CODE REVIEW

File: tests/Support/NovaTest/NovaTestMockTrait.php [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+trait NovaTestMockTrait
+{
+    use GeneratorMockTrait, MockClassTrait;
+
+    public function mockNovaResourceTestGenerator(): void
+    {
+        $this->mockClass(NovaTestGenerator::class, [
+            [
+                'method' => 'getModelFields',
```
Comment (DenTray) [inline tests/Support/NovaTest/NovaTestMockTrait.php:18]: let's mock only methods which result can not been changed from the third-party factors, e.g. mocking `file_exists` method is good.

For this case - I belive we should mock the `getFields` fethod of called model, or make a test model with the `getFields` methods

**Review by DenTray — CHANGES_REQUESTED** [review]

File: tests/Support/Shared/GeneratorMockTrait.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+namespace RonasIT\Support\Tests\Support\Shared;
+
+use Illuminate\Support\Facades\View;
+use phpmock\Mock;
+use phpmock\MockBuilder;
+use PHPUnit\Framework\MockObject\MockObject;
+
+trait GeneratorMockTrait
+{
+    protected function mockClassWithReturn($className, $methods = [], $disableConstructor = false): MockObject
```
Comment (DenTray) [inline tests/Support/Shared/GeneratorMockTrait.php:12]: let's move similar method from the laravel helpers package and use it here

**Review by DenTray — CHANGES_REQUESTED** [review]

**Comment by DenTray** [comment]
@KonstantinLapkovsky please check all other mock traits according my comments

File: tests/fixtures/ControllerGeneratorTest/api.php [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+use App\Controllers\PostController;
+
```
[the remark this answers is not in the document]
Comment (KonstantinLapkovsky) [reply tests/fixtures/ControllerGeneratorTest/api.php:4]: @DenTray эта пустая строчка нужна, чтобы отделять новую группу реквестов от предыдущих существующих. Две пустые строки будут только если мы добавляем роуты в пустой `api.php` файл.

File: tests/Support/NovaTest/NovaTestMockTrait.php (hunk 1 above)
Comment (KonstantinLapkovsky) [reply tests/Support/NovaTest/NovaTestMockTrait.php:18]: @DenTray the logic for getting model fields is complicated and will be covered in different test cases in a future request. For now, I include in the test minimum logic to assert only generator results.

File: tests/Support/Shared/GeneratorMockTrait.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        View::addNamespace('entity-generator', getcwd() . '/stubs');
+    }
+
+    public function mockClassExistsFunction(bool $result = true): Mock
+    {
+        return $this->mockNativeFunction('\\RonasIT\\Support\\Generators', 'class_exists', $result);
+    }
+
+    public function mockCheckingNonExistentNovaPackageExistence(): Mock
```
Comment (DenTray) [inline tests/Support/Shared/GeneratorMockTrait.php:53]:
```suggestion
    public function mockCheckingNovaPackageExistence(bool $result = false): Mock
```

File: tests/Support/Controller/ControllerMockTrait.php [hunk 5]
```diff
@@ -0,0 +1,75 @@
+<?php
+
+namespace RonasIT\Support\Tests\Support\Controller;
```
Comment (DenTray) [inline tests/Support/Controller/ControllerMockTrait.php:3]:
```suggestion
namespace RonasIT\Support\Tests\Support\ControllerGeneratorTest;
```

File: tests/Support/Shared/GeneratorMockTrait.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
+<?php
+
+namespace RonasIT\Support\Tests\Support\Shared;
```
Comment (DenTray) [inline tests/Support/Shared/GeneratorMockTrait.php:3]:
```suggestion
namespace RonasIT\Support\Tests\Support;
```

File: tests/Support/NovaResource/NovaResourceMockTrait.php [hunk 7]
```diff
@@ -0,0 +1,74 @@
+<?php
+
+namespace RonasIT\Support\Tests\Support\NovaResource;
```
Comment (DenTray) [inline tests/Support/NovaResource/NovaResourceMockTrait.php:3]:
```suggestion
namespace RonasIT\Support\Tests\Support\NovaResourceGeneratorTest;
```

File: tests/Support/NovaTest/NovaTestMockTrait.php [hunk 8]
```diff
@@ -0,0 +1,126 @@
+<?php
+
+namespace RonasIT\Support\Tests\Support\NovaTest;
```
Comment (DenTray) [inline tests/Support/NovaTest/NovaTestMockTrait.php:3]:
```suggestion
namespace RonasIT\Support\Tests\Support\NovaTestGeneratorTest;
```

File: tests/Support/Shared/GeneratorMockTrait.php [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        return $this->mockNativeFunction('\\RonasIT\\Support\\Generators', 'class_exists', $result);
+    }
+
+    public function mockCheckingNonExistentNovaPackageExistence(): Mock
+    {
+        return $this->mockClassExistsFunction(false);
+    }
+
+    public function classExistsMethodCall(?string $path, ?string $className, bool $result = true): array
```
Comment (DenTray) [inline tests/Support/Shared/GeneratorMockTrait.php:58]: It looks like optional args are using only in the `mockResourceGeneratorForNonExistingNovaResource` method. Could we use filled args there and remove optional here?

**Review by DenTray — CHANGES_REQUESTED** [review]

File: tests/Support/GeneratorMockTrait.php [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+namespace RonasIT\Support\Tests\Support;
+
+use Illuminate\Support\Facades\View;
+use phpmock\Mock;
+use phpmock\MockBuilder;
+
+trait GeneratorMockTrait
+{
+    public function mockNativeFunction(string $namespace, string $name, $result): Mock
```
Comment (DenTray) [inline tests/Support/GeneratorMockTrait.php:11]: please check that such logic already exists in the laravel helpers package and use it

**Review by DenTray — CHANGES_REQUESTED** [review]

File: tests/Support/ControllerGeneratorTest/ControllerGeneratorMockTrait.php [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+trait ControllerGeneratorMockTrait
+{
+    use GeneratorMockTrait;
+    use MockTrait;
+
+    public function mockFilesystemWithoutRoutesFile(): void
+    {
+        $fileSystemMock = new FileSystemMock;
+        $fileSystemMock->services = [
+            'PostService.php' => '<?php'
```
Comment (DenTray) [inline tests/Support/ControllerGeneratorTest/ControllerGeneratorMockTrait.php:18]:
```suggestion
            'PostService.php' => $this->mockPhpFileContent(),
```

File: tests/Support/FileSystemMock.php [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    public $testClasses = null;
+    public $routes = null;
+
+    public function setStructure()
+    {
+        $structure = ['app' => []];
+
+        if (!is_null($this->novaModels)) {
+            $structure['app']['Nova'] = [];
+            foreach ($this->novaModels as $novaModel => $content) {
```
Comment (DenTray) [inline tests/Support/FileSystemMock.php:25]:
```suggestion

            foreach ($this->novaModels as $novaModel => $content) {
```

**Review by DenTray — CHANGES_REQUESTED** [review]

File: tests/Support/GeneratorMockTrait.php [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+trait GeneratorMockTrait
+{
+    public function mockClassExistsFunction(string $className, bool $result = true, bool $autoloadArg = true): void
+    {
+        $this->mockNativeFunction('\RonasIT\Support\Generators', [
+            $this->functionCall(
+                name: 'class_exists',
+                arguments: [$className, $autoloadArg],
+                result: $result,
+            ),
```
Comment (DenTray) [inline tests/Support/GeneratorMockTrait.php:16]:
```suggestion
            $this->classExistsMethodCall($className, $result, $autoloadArg),
```

File: tests/Support/NovaTestGeneratorTest/NovaTestGeneratorMockTrait.php [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+use RonasIT\Support\Tests\Support\FileSystemMock;
+use RonasIT\Support\Tests\Support\GeneratorMockTrait;
+use RonasIT\Support\Traits\MockTrait;
+
+trait NovaTestGeneratorMockTrait
+{
+    use GeneratorMockTrait;
+    use MockTrait;
+
+    public function mockNovaResourceTestGenerator(): void
```
Comment (DenTray) [inline tests/Support/NovaTestGeneratorTest/NovaTestGeneratorMockTrait.php:15]: could we reduce of this method and run generation on fake test model?

**Review by DenTray — CHANGES_REQUESTED** [review]

File: tests/NovaResourceGeneratorTest.php [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+        });
+    }
+
+    public function testCreateNovaResourceWithMissingModel()
+    {
+        $this->mockNovaServiceProviderExists();
+
+        $this->expectException(ClassNotExistsException::class);
+        $this->expectExceptionMessage('Cannot create Post Nova resource cause Post Model does not exists. '
+            . "Create a Post Model by himself or run command 'php artisan make:entity Post --only-model'");
```
[the remark this answers is not in the document]
Comment (DenTray) [reply tests/NovaResourceGeneratorTest.php:49]: could we use already implemented `assertExceptionThrowed` helper instead?

File: tests/NovaTestGeneratorTest.php [hunk 16]
```diff
@@ -2,74 +2,87 @@
...
 use RonasIT\Support\Generators\NovaTestGenerator;
-use RonasIT\Support\Tests\Support\NovaTestMockTrait;
+use RonasIT\Support\Tests\Support\NovaTestGeneratorTest\NovaTestGeneratorMockTrait;
 
 class NovaTestGeneratorTest extends TestCase
 {
-    use NovaTestMockTrait;
+    use NovaTestGeneratorMockTrait;
 
     public function testCreateNovaTestsResourceNotExists()
```
Comment (DenTray) [inline tests/NovaTestGeneratorTest.php:16]:
```suggestion
    public function testGenerateResourceNotExists()
```

File: tests/NovaTestGeneratorTest.php [hunk 17]
```diff
@@ -2,74 +2,87 @@
...
-                ->generate();
-        } finally {
-            $mock->disable();
-        }
+        app(NovaTestGenerator::class)
+            ->setModel('Post')
+            ->generate();
     }
 
     public function testCreateNovaTestAlreadyExists()
```
Comment (DenTray) [inline tests/NovaTestGeneratorTest.php:36]:
```suggestion
    public function testGenerateNovaTestAlreadyExists()
```

File: tests/NovaTestGeneratorTest.php [hunk 18]
```diff
@@ -2,74 +2,87 @@
...
-        $this->rollbackToDefaultBasePath();
-
         $this->assertGeneratedFileEquals('created_resource_test.php', 'tests/NovaWelcomeBonusTest.php');
         $this->assertGeneratedFileEquals('dump.sql', 'tests/fixtures/NovaWelcomeBonusTest/nova_welcome_bonus_dump.sql');
         $this->assertGeneratedFileEquals('create_welcome_bonus_request.json', 'tests/fixtures/NovaWelcomeBonusTest/create_welcome_bonus_request.json');
         $this->assertGeneratedFileEquals('create_welcome_bonus_response.json', 'tests/fixtures/NovaWelcomeBonusTest/create_welcome_bonus_response.json');
         $this->assertGeneratedFileEquals('update_welcome_bonus_request.json', 'tests/fixtures/NovaWelcomeBonusTest/update_welcome_bonus_request.json');
+    }
+
+    public function testCreateWithMissingNovaPackage()
```
Comment (DenTray) [inline tests/NovaTestGeneratorTest.php:73]:
```suggestion
    public function testGenerateNovaPackageNotInstall()
```

File: tests/Support/GeneratorMockTrait.php [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+        ]);
+    }
+
+    public function mockCheckingNovaPackageExistence(string $className, bool $result = false): void
+    {
+        $this->mockClassExistsFunction($className, $result);
+    }
```
Comment (DenTray) [inline tests/Support/GeneratorMockTrait.php:23]: seems this method is not using anymore

```suggestion
```

File: src/Generators/NovaResourceGenerator.php [hunk 20]
```diff
@@ -112,7 +112,7 @@ protected function prepareNovaFields(): array
         return $result;
     }
 
-    protected function getFieldsForCreation(): array
+    public function getFieldsForCreation(): array
```
Comment (DenTray) [inline src/Generators/NovaResourceGenerator.php:115]: should we change visibility for this function?

**Review by DenTray — CHANGES_REQUESTED** [review]

File: tests/NovaResourceGeneratorTest.php [hunk 21]
```diff
@@ -0,0 +1,119 @@
+<?php
+
+namespace RonasIT\Support\Tests;
+
+use Doctrine\DBAL\Schema\Column;
+use Doctrine\DBAL\Types\DateTimeType;
+use Doctrine\DBAL\Types\IntegerType;
+use Doctrine\DBAL\Types\StringType;
+use Illuminate\Support\Facades\Event;
+use ReflectionClass;
```
Comment (DenTray) [inline tests/NovaResourceGeneratorTest.php:10]:
```suggestion
```

**Review by DenTray — CHANGES_REQUESTED** [review]

File: src/Generators/AbstractTestsGenerator.php [hunk 22]
```diff
@@ -184,6 +184,7 @@ protected function getMockModel($model): array
         $hasFactory = method_exists($modelClass, 'factory') && $this->classExists('factory', "{$model}Factory");
         $factory = ($hasFactory) ? $modelClass::factory() : factory($modelClass);
 
+        dd($factory);
```
Comment (DenTray) [inline src/Generators/AbstractTestsGenerator.php:187]:
```suggestion
```

**Review by DenTray — CHANGES_REQUESTED** [review]

File: tests/Support/NovaTestGeneratorTest/NovaTestGeneratorMockTrait.php [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
```
Comment (DenTray) [inline tests/Support/NovaTestGeneratorTest/NovaTestGeneratorMockTrait.php:15]:
```suggestion
    public function mockNovaRequestClassCall(): void
```

File: tests/NovaTestGeneratorTest.php [hunk 24]
```diff
@@ -2,74 +2,87 @@
...
-    public function testCreateWithActions()
+    public function testSuccess()
     {
-        $functionMock = $this->mockClassExistsFunction();
+        $this->mockNovaServiceProviderExists();
 
         $this->mockFilesystem();
-        $this->setupConfigurations();
-        $this->mockViewsNamespace();
         $this->mockNovaResourceTestGenerator();
```
Comment (DenTray) [inline tests/NovaTestGeneratorTest.php:60]:
```suggestion
        $this->mockNovaRequestClassCall();
```

**Review by DenTray — APPROVED** [review]

**Review by DenTray — APPROVED** [review]
